File: Controllers/UserController.js
const chalk = require("chalk");
const admin = require("firebase-admin");
const { Router } = require("express");
const { UserRepository } = require("../Repository/UserRepository");
const { UserBusiness } = require("../Business/UserBusiness");
class UserController {
  /**
   * Returns a router instance representing the UserController
   * @returns {Router}
   */
  static getRouter() {
    const router = Router();
    router.get("/details", UserController.GetUserInformationAsync);
    router.post("/", UserController.CreateUserAsync);
    router.get("/", UserController.GetUsersAsync);
    router.delete("/:uid", UserController.DeleteUserAsync);
    return router;
  }

  /**
   * Returns information about the requesting user.
   * @param {Request} req
   * @param {Response} res
   */
  static async GetUserInformationAsync(req, res) {
    const user = {
      uid: req.userData.uid,
      isAdmin: req.userData.isAdmin,
      name: req.userData.name
    };

    res.json(user);
  }

  /**
   * Returns a list of users.
   * @param {Request} req
   * @param {Response} res
   */
  static async GetUsersAsync(req, res) {
    if (!req.userData.isAdmin) {
      res.status(401).send();
      console.log(chalk.yellow("Unauthorized attempt to get user data."));
      return;
    }

    const users = await UserRepository.GetAllUsersAsync();
    const result = users.map((user) => {
      return {
        uid: user.uid,
        name: user.name,
        isAdmin: user.isAdmin,
        department: user.department,
        email: user.email
      };
    });
    res.json(result);
  }

  /**
   * Creates a new user.
   * @param {Request} req
   * @param {Response} res
   */
  static async CreateUserAsync(req, res) {
    if (!req.userData.isAdmin) {
      res.status(401).send();
      console.log(chalk.yellow("Unauthorized attempt to create a new user.\n", req.body));
      return;
    }

    try {
      const createUserResult = await UserBusiness.CreateUserAsync(req.body);
      console.log(createUserResult);

      if (createUserResult.error) {
        res.status(400).json({ reason: createUserResult.reason });
        return;
      }

      res.json(createUserResult.user);
    } catch (error) {
      console.log(chalk.yellow("Unexpected error occured while creating new user:\n", error));
      res.status(500).send();
    }
  }

  /**
   * Deletes a user
   * @param {Request} req
   * @param {Response} res
   */
  static async DeleteUserAsync(req, res) {
    if (!req.userData.isAdmin) {
      res.status(401).send();
      console.log(chalk.yellow(`Unauthorized attempt to delete user ${req.params.uid}.\n`, req.body));
      return;
    }

    try {
      await admin.auth().deleteUser(req.params.uid);
      await UserRepository.DeleteUserAsync(req.params.uid);
      res.status(204).send();
    } catch (error) {
      console.log(chalk.yellow("Unexpected error occured while creating new user:\n", error));
      res.status(500).send();
    }
  }
}

module.exports = { UserController };

File: src/ChatBot.Service/Controllers/UserController.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using ChatBot.AuthProvider.Contract.Exceptions;
using ChatBot.Business.Contracts.User;
using ChatBot.Business.Contracts.User.Exceptions;
using ChatBot.Repository.Contracts;
using ChatBot.Service.Mappers;
using ChatBot.Service.Models;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Microsoft.Extensions.Logging;
using Shouldly;

namespace ChatBot.Service.Controllers
{
    /// <summary>
    ///     Provided methods to manage users
    /// </summary>
    [Route("api/[controller]")]
    [ApiController]
    public class UserController : ControllerBase
    {
        private readonly ILogger<UserController> _logger;
        private readonly IUserBusiness _userBusiness;
        private readonly IUserRepository _userRepository;

        public UserController(IUserRepository userRepository, ILogger<UserController> logger,
            IUserBusiness userBusiness)
        {
            userRepository.ShouldNotBeNull();
            logger.ShouldNotBeNull();
            userBusiness.ShouldNotBeNull();

            _userRepository = userRepository;
            _logger = logger;
            _userBusiness = userBusiness;
        }

        /// <summary>
        ///     Returns details about the user that is currently signed in
        /// </summary>
        /// <returns></returns>
        [HttpGet("/details")]
        [Authorize]
        [ProducesResponseType(typeof(UserDetails), 200)]
        [ProducesResponseType(400)]
        [ProducesResponseType(500)]
        public async Task<IActionResult> GetUserDataAsync()
        {
            try
            {
                _logger.LogInformation("Getting user details");
                var userId = GetCurrentUserId();

                if (userId == null)
                {
                    _logger.LogWarning("Could not determine userid from token");
                    return BadRequest();
                }


                var userData = await _userRepository.GetUserDetailsAsync(userId);

                if (userData != null) return Ok(userData);

                _logger.LogWarning($"Could not find data for uid {userId}.");
                return NotFound();
            }
            catch (Exception ex)
            {
                _logger.LogError("Unexpected error occured while getting user details", ex);
                return StatusCode(500);
            }
        }


        /// <summary>
        ///     Creates a new user
        /// </summary>
        /// <param name="user"></param>
        /// <returns></returns>
        [HttpPost]
        [Authorize]
        [ProducesResponseType(typeof(CreateUserResponse), 200)]
        [ProducesResponseType(400)]
        [ProducesResponseType(500)]
        public async Task<IActionResult> CreateUserAsync([FromBody] CreateUserRequest user)
        {
            try
            {
                _logger.LogInformation("Trying to create a new user.");

                var currentUserId = GetCurrentUserId();
                var isRequestingUserAdmin = await _userBusiness
                    .CheckAdminPrivilegesAsync(currentUserId)
                    .ConfigureAwait(false);

                if (!isRequestingUserAdmin)
                {
                    _logger.LogInformation($"User with id {currentUserId} tried to manipulate user data.");
                    return Unauthorized();
                }

                var newUserModel = await _userBusiness
                    .CreateUserAsync(user.Map())
                    .ConfigureAwait(false);

                return Ok(newUserModel.Map());
            }
            catch (InvalidUserDataException ex)
            {
                _logger.LogWarning(ex, "The specified object contained errors", user);
                return BadRequest();
            }
            catch (UserAlreadyExistsException ex)
            {
                _logger.LogWarning("The user that was tried to create already exists", ex);
                return BadRequest();
            }
            catch (Exception ex)
            {
                _logger.LogError("An unexpected error occured while creating a new user.", ex);
                return StatusCode(500);
            }
        }


        /// <summary>
        ///     Deletes the user specified in the route
        /// </summary>
        /// <param name="userId"></param>
        /// <returns></returns>
        [HttpDelete("{userId}")]
        [Authorize]
        [ProducesResponseType(204)]
        [ProducesResponseType(401)]
        [ProducesResponseType(500)]
        public async Task<IActionResult> DeleteUserAsync([FromRoute] string userId)
        {
            try
            {
                _logger.LogInformation("Trying to delete user");

                var currentUserId = GetCurrentUserId();
                var isRequestingUserAdmin = await _userBusiness
                    .CheckAdminPrivilegesAsync(currentUserId)
                    .ConfigureAwait(false);

                if (!isRequestingUserAdmin)
                {
                    _logger.LogInformation($"User with id {currentUserId} tried to manipulate user data.");
                    return Unauthorized();
                }

                var userData = await _userRepository
                    .GetAllUsersAsync()
                    .ConfigureAwait(false);

                return Ok(userData);
            }
            catch (ShouldAssertException)
            {
                _logger.LogInformation("No userId was specified");
                return NotFound();
            }
            catch (Exception ex)
            {
                _logger.LogError(ex,"Unexpected error occured while deleting a user", userId);
                return StatusCode(500);
            }
        }

        /// <summary>
        ///     Returns a list of all registered users
        /// </summary>
        /// <returns></returns>
        [HttpGet]
        [Authorize]
        [ProducesResponseType(typeof(IEnumerable<UserDetails>), 200)]
        [ProducesResponseType(401)]
        [ProducesResponseType(500)]
        public async Task<IActionResult> GetAllUsersAsync()
        {
            try
            {
                _logger.LogInformation("Trying to get all users data");

                var currentUserId = GetCurrentUserId();
                var isRequestingUserAdmin = await _userBusiness
                    .CheckAdminPrivilegesAsync(currentUserId)
                    .ConfigureAwait(false);

                if (!isRequestingUserAdmin)
                {
                    _logger.LogInformation($"User with id {currentUserId} tried to manipulate user data.");
                    return Unauthorized();
                }

                var allUsers = await _userRepository
                    .GetAllUsersAsync()
                    .ConfigureAwait(false);

                return Ok(allUsers);
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Unexpected error occured while getting all users");
                return StatusCode(500);
            }
        }

        /// <summary>
        ///     Returns the id of the current user id
        /// </summary>
        /// <returns></returns>
        private string GetCurrentUserId()
        {
            var userId = User.Claims.FirstOrDefault(claim => claim.Type == "user_id")?.Value;
            userId.ShouldNotBeNullOrWhiteSpace();
            return userId;
        }
    }
}